### PR TITLE
Async allocator interface

### PR DESCRIPTION
**Pull request by idegtiarenko**
Wire listeners from create index operation into desired balance computation.
This pr ignores the other flows for now and uses a noop listener. 
They could be implemented in a followups.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java [hunk 1]
```diff
@@ -67,12 +73,12 @@ public void allocate(RoutingAllocation allocation) {
         desiredBalanceComputation.onNewInput(
             new DesiredBalanceInput(allocation.immutableClone(), new ArrayList<>(allocation.routingNodes().unassigned().ignored()))
         );
+        listener.onResponse(AcknowledgedResponse.TRUE);// TODO listener need to be passed to the above call
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java:76]: Could be done after https://github.com/elastic/elasticsearch/pull/86217 is merged

Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java:76]: It looks like it is not enough to complete this listener once `desiredBalanceComputation` is complete as `processInput` might result in reroute that is async itself.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 2]
```diff
@@ -285,13 +287,20 @@ public void createIndex(final CreateIndexClusterStateUpdateRequest request, fina
...
+
+        var future = new ListenableFuture<Void>();
+
         submitUnbatchedTask(
             "create-index [" + request.index() + "], cause [" + request.cause() + "]",
-            new AckedClusterStateUpdateTask(Priority.URGENT, request, listener) {
+            new AckedClusterStateUpdateTask(
+                Priority.URGENT,
+                request,
+                future.delegateFailure((delegate, response) -> future.addListener(listener.map(ignored -> response)))
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:298]: Should this be `listener.delegateFailure`? We want to complete `listener` on a failure, we don't care about `future`.

File: test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java [hunk 3]
```diff
@@ -2442,10 +2440,4 @@ public Settings onNodeStopped(String nodeName) throws Exception {
             }
         });
     }
-
-    protected static void awaitDesiredBalanceShardsAllocator() throws Exception {
-        if (internalCluster().getCurrentMasterNodeInstance(ShardsAllocator.class)instanceof DesiredBalanceShardsAllocator dbsa) {
-            assertBusy(() -> assertTrue(dbsa.isIdle()));
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:2448]: I think this is the only caller of `isIdle` so we can remove it and the flag on which it depends.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 4]
```diff
@@ -508,7 +522,8 @@ private ClusterState applyCreateIndexRequestWithV1Templates(
         final CreateIndexClusterStateUpdateRequest request,
         final boolean silent,
         final List<IndexTemplateMetadata> templates,
-        final BiConsumer<Metadata.Builder, IndexMetadata> metadataTransformer
+        final BiConsumer<Metadata.Builder, IndexMetadata> metadataTransformer,
+        final ActionListener<Void> listener
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:526]: Could we call this `rerouteListener` (here and elsewhere in this PR) just to clarify what it's listening for?

**Review by DaveCTurner — COMMENTED** [review]
Looks good, one change needed and a couple of other suggestions.

File: test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java (hunk 3 above)
Comment (idegtiarenko) [reply test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:2448]: Unfortunately it is still used in `org.elasticsearch.cluster.coordination.Coordinator#hasIdleJoinValidationService` and `org.elasticsearch.cluster.coordination.JoinValidationServiceTests#testConcurrentBehaviour`

Comment (DaveCTurner) [reply test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:2448]: ah ok

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**